### PR TITLE
Add keybinding & setting to perform cleanup actions manually. 

### PR DESCRIPTION
**Pull request by hopehadfield**
Implements request in #3424.

Requires https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3010.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1536,6 +1547,11 @@
         "key": "ctrl+shift+v",
         "mac": "cmd+shift+v",
         "when": "explorerViewletFocus && config.editor.pasteAs.enabled"
+      },
+      {
+        "command": "java.action.doCleanup",
+        "key": "ctrl+shift+alt+s",
+        "when": "javaLSReady && config.java.cleanup.manualKey && editorLangId == java"
```
Comment (rgrunber) [inline package.json:1554]: Is `config.java.cleanup.manualKey` needed ? I don't see any such setting to test.

File: src/sourceAction.ts [hunk 2]
```diff
@@ -67,6 +87,17 @@ function registerOverrideMethodsCommand(languageClient: LanguageClient, context:
     }));
 }
 
+function registerManualCleanupCommand(languageClient: LanguageClient, context: ExtensionContext): void {
+    context.subscriptions.push(commands.registerCommand(Commands.MANUAL_CLEANUP, async () => {
+        const languageClient: LanguageClient | undefined = await getActiveLanguageClient();
+        const document = window.activeTextEditor !== undefined ? window.activeTextEditor.document : undefined;
+        if (document !== undefined){
```
Comment (rgrunber) [inline src/sourceAction.ts:94]: This is just an extra check for safety right ? I would assume the document should always exist if you had  something like `editorTextFocus && editorLangId == java` in the when clause ? Is it worth adding `editorTextFocus` above ?

File: package.json [hunk 3]
```diff
@@ -1189,6 +1190,11 @@
           "scope": "window",
           "order": 10
         },
+        "java.saveActions.enabled": {
+          "type": "boolean",
+          "default": true,
+          "description": "Enable/disable key for manual cleanup actions typically done on save"
```
Comment (rgrunber) [inline package.json:1196]: I think the description is wrong here. Applying the cleanups through a keyboard should never be guarded by a setting. The ability to perform the cleanups on save is what's guarded now.

File: src/protocol.ts [hunk 4]
```diff
@@ -233,6 +233,10 @@ export namespace OrganizeImportsRequest {
     export const type = new RequestType<CodeActionParams, WorkspaceEdit, void>('java/organizeImports');
 }
 
+export namespace ManualCleanupRequest {
+    export const type = new RequestType<TextDocumentIdentifier, WorkspaceEdit, void>('java/manualCleanup');
```
Comment (rgrunber) [inline src/protocol.ts:237]: I think I would name it just `java/cleanup`.

File: package.json (hunk 1 above)
Comment (hopehadfield) [reply package.json:1554]: Leftover from when I had a setting that enabled/disabled the key (as is the description below), just realized I forgot to push the change. Will update.

File: src/sourceAction.ts (hunk 2 above)
Comment (hopehadfield) [reply src/sourceAction.ts:94]: I had this for safety prior to having `editorLangId == java` in the when clause. That condition alone is sufficient for ensuring the document exists. I'll remove the extra check.

Comment (rgrunber) [reply src/sourceAction.ts]: Ok. I would maybe verify whether `editorTextFocus && editorLangId == java"` gives us anything extra. I see it used in a few spots. I think editorLangId is just about the editor that is visible, but editorTextFocus asserts the cursor is also in the editor. Maybe we would want the cleanup to only work when the cursor is actually in the editor.

File: src/sourceAction.ts [hunk 5]
```diff
@@ -67,6 +87,14 @@ function registerOverrideMethodsCommand(languageClient: LanguageClient, context:
     }));
 }
 
+function registerCleanupCommand(languageClient: LanguageClient, context: ExtensionContext): void {
+    context.subscriptions.push(commands.registerCommand(Commands.MANUAL_CLEANUP, async () => {
+        const languageClient: LanguageClient | undefined = await getActiveLanguageClient();
+        const workspaceEdit = await languageClient.sendRequest(CleanupRequest.type, languageClient.code2ProtocolConverter.asTextDocumentIdentifier(window.activeTextEditor.document));
```
Comment (rgrunber) [inline src/sourceAction.ts:93]: I would just add a comment here to indicate that `MANUAL_CLEANUP` is only active when `editorlangId == java`.

File: package.json [hunk 6]
```diff
@@ -54,7 +54,8 @@
     "workspaceContains:.classpath",
     "workspaceContains:*/.classpath",
     "onCommand:_java.templateVariables",
-    "onCommand:_java.metadataFilesGeneration"
+    "onCommand:_java.metadataFilesGeneration",
+    "onCommand:java.action.filesExplorerPasteActionNoLS"
```
Comment (rgrunber) [inline package.json:58]: Was this from an experiment ?

**Review by rgrunber — APPROVED** [review]
Overall looks good. Just some minor things. I think this part is ready.

File: package.nls.json [hunk 7]
```diff
@@ -26,5 +26,6 @@
 	"java.clean.sharedIndexes": "Clean Shared Indexes",
 	"java.server.restart": "Restart Java Language Server",
 	"java.edit.smartSemicolonDetection": "Java Smart Semicolon Detection",
-	"java.action.filesExplorerPasteAction": "Paste clipboard text into a file"
+	"java.action.filesExplorerPasteAction": "Paste clipboard text into a file",
+	"java.action.doCleanup": "Performs cleanup actions"
```
Comment (rgrunber) [inline package.nls.json:30]: I noticed most of the labels for key bindings capitalize the start of each word so I would do the same here.

File: package.json (hunk 6 above)
Comment (hopehadfield) [reply package.json]: Honestly not sure where that came from, I missed this. Must have been some old stash from when I was doing the paste keybinding? Got rid of it.

File: package.json [hunk 8]
```diff
@@ -1236,6 +1236,34 @@
           "scope": "window",
           "order": 10
         },
+        "java.cleanup.actionsOnSave": {
+          "type": "array",
+          "markdownDeprecationDescription": "**Deprecated, please use 'java.clenaup.actions' instead.**",
```
Comment (rgrunber) [inline package.json:1241]: This doesn't exist. There's either `markdownDeprecationMessage` or `deprecationMessage`. You can check this by typing out the property in a settings file and there should be a warning showing the deprecation message. I would go with just `deprecationMessage` though like some other instances.

**Pull request merged**